modules/images/image-loading-optimization/storage/post-type.php

```diff
@@ -110,7 +110,40 @@ function ilo_parse_stored_url_metrics( WP_Post $post ): array {
 		);
 		$url_metrics = array();
 	}
-	return $url_metrics;
+
+	return array_values(
+		array_filter(
+			$url_metrics,
+			static function ( $url_metric ) use ( $trigger_error ) {
+				// TODO: If we wanted, we could use the JSON Schema to validate the stored metrics.
+				$is_valid = (
+					is_array( $url_metric )
+					&&
+					isset(
+						$url_metric['viewport']['width'],
+						$url_metric['viewport']['height'],
+						$url_metric['elements']
+					)
+					&&
+					is_int( $url_metric['viewport']['width'] )
+					&&
+					is_array( $url_metric['elements'] )
+				);
+
+				if ( ! $is_valid ) {
+					$trigger_error(
+						sprintf(
+							/* translators: %s is post type slug */
+							__( 'Unexpected shape to JSON array in post_content of %s post type.', 'performance-lab' ),
+							ILO_URL_METRICS_POST_TYPE
+						)
+					);
+				}
+
+				return $is_valid;
+			}
+		)
+	);
```

Comment:
Nevertheless, the data would be stored as JSON in the database and not as a serialized class. So this validation would still have to be done somewhere.

Reply:
Right, I think it could then be done in the class. I would expect a class to e.g. receive this associative array as constructor parameter and then validate it.